### PR TITLE
fix(engine): key a constructor on the class that declares it

### PR DESCRIPTION
**Pull request by Svilen-Stefanov**
First of three. Prerequisite for #544, the Java rename.

`class_to_ctors` found a constructor's class by slicing the qualified name at its first `(`:

```python
paren_idx = qname.find("(")
class_qname = qname[:paren_idx]
```

That names the real class only where the naming scheme happens to double it, which today is **Java alone**. `core/Animal.java` gives the class `src.main.java.core.Animal.Animal` because the file stem is folded in twice, and its constructor `…Animal.Animal(String)` slices back to exactly that. Every other language indexes nothing: C# collapses the stem so the slice overshoots by a segment, TypeScript's `constructor` has no parentheses, and Python reports `__init__` as a method.

Keying on the declaring symbol recorded during the walk says what was meant, and it cannot pick a dual-registration alias, which the slice could.

## Two gates on expansion

**Which languages expand at all.** Correct keying makes the index resolve for C# and TypeScript too, where it never had. Whether a class edge *should* fan out into that class's constructors is a different question from what a constructor is called — it is right only where the server resolves `new Foo()` to the type rather than to the constructor. So expansion is an adapter property, set for Java, and **the graph is unchanged in every language**.

**Which call sites count as construction — the source decides.** A definition query resolves

```java
public String purr() {
    return getName() + " purrs";   // Cat.java:36
}
```

to both the inherited `Animal.getName()` *and* its declaring class `Animal`. The class half was then expanded into a constructor call, so the graph claimed `Cat.purr() -> Animal.Animal(String)` at a line that runs no constructor.

`SourceInspector` already parses the file and already knows every shape of construction — `new X()`, C#'s target-typed `new(...)` and `: base(...)`, Java's `super(...)`, an applied attribute, and `Dog::new` as against `Dog::speak`. It now answers `is_construction_site(site)` directly, populated **in the same tree walk that builds the usage index**, so it costs no extra parsing and reuses the per-file cache. No qualified-name parsing and no graph inference are involved.

Expansion additionally skips a site where the server has already named a constructor itself: where jdtls resolved `new Dog("Rex", 5)` to `Dog(String, int)`, fanning the class edge out would add every *other* overload beside it. The question is per site, because one caller can construct on one line and call a member on the next.

## Verified against the real engine

On `java_edge_cases_project`, this PR leaves the graph **identical to `main`** — 94 references and 86 edges, the same sets.

It becomes load-bearing under #544, where the recovered class edges would otherwise invent **nine** constructor calls from methods that construct nothing. With it: 110 edges, 0 lost, 24 gained, 0 invented — which is what CodeBoarding-tests#27 pins.

## Verified

`pytest tests/` → 2258 passed on the stack tip (3 pre-existing `test_cli_parser` failures deselected; they fail identically on a clean `main`). `black`, `mypy` clean. 18/18 C# and Java edge-case integration tests pass against the real engine.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved call-graph accuracy for Java constructors and object creation.
  * Prevented member access, non-constructor symbols, and aliases from being incorrectly treated as constructor calls.
  * Improved ownership tracking for nested symbols and constructors.
  * Added more reliable detection of constructor usage across supported languages.
  * Disabled constructor expansion for languages that do not require it.

* **Tests**
  * Added coverage for constructor edge expansion, construction detection, and symbol indexing behavior.

### CODE REVIEW

File: static_analyzer/engine/language_adapter.py [hunk 1]
```diff
@@ -356,6 +356,16 @@ def resolves_collection_initializers(self) -> bool:
         once per element, rather than only constructing the type."""
         return False
 
+    @property
+    def expands_constructors(self) -> bool:
+        """Whether an edge to a class should also reach that class's constructors.
+
+        Why: only for servers that resolve ``new Foo()`` to the type rather than to the
+        constructor. Where the server already names the constructor, expanding invents a
+        call for every plain mention of the type.
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/language_adapter.py]: Fixed — down to one line plus a short `Why:`.

File: static_analyzer/engine/models.py [hunk 2]
```diff
@@ -26,6 +26,10 @@ class SymbolInfo:
     # Promoted to CLASS for having callable children: a namespace object, or a
     # `const raw = await fn(() => ...)` wrapper that is a scope rather than a name.
     promoted_from_variable: bool = False
+    # Qualified name of the declaring symbol, empty for a top-level symbol and for every
+    # alias. Why: slicing the qualified name to find the owner agrees with the real one only
+    # by luck of the naming scheme.
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/models.py]: Fixed — two lines, the second a `Why:`.

File: static_analyzer/engine/symbol_table.py [hunk 3]
```diff
@@ -183,14 +185,13 @@ def build_indices(self) -> None:
...
-            if sym.kind == NodeType.CONSTRUCTOR:
-                # Find the class this constructor belongs to by stripping the params
-                paren_idx = qname.find("(")
-                if paren_idx != -1:
-                    class_qname = qname[:paren_idx]
-                    self._class_to_ctors.setdefault(class_qname, []).append(qname)
+        # Class -> constructors, keyed on the symbol that declares them. Why not a slice of
+        # the qualified name at its first "(": that names the real class only where the
+        # naming scheme happens to double it, and it cannot tell a primary symbol from a
+        # dual-registration alias.
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/symbol_table.py]: Fixed — three lines down to two.

File: static_analyzer/engine/call_graph_builder.py [hunk 4]
```diff
@@ -303,18 +303,39 @@ def _postprocess_edges(self, edge_set: EdgeMap) -> EdgeMap:
...
+            # A site where the caller also reaches inside the class is a member access, not a
+            # construction: `return getName() + "!"` resolves to both the method and its
+            # declaring class, and expanding there invents a constructor call for a line that
+            # runs none. Judged per site, because one caller can do both -- construct on one
+            # line and call a member on the next.
+            member_sites = {
+                (site.file, site.line, site.column)
+                for (other_src, other_dst), sites in edge_set.items()
+                if other_src == src and other_dst.startswith(f"{dst}.")
+                for site in sites
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/call_graph_builder.py]: Good catch, and it was worse than quadratic time: I was also comparing with `startswith`, so a destination whose parameter list contains a dot (`feed(java.util.List)`) could match the wrong owner.

Now built once, before the loop, as `(source, enclosing scope) -> sites`. The destination is sliced at its first `(` so a parameter list cannot contribute segments, and every enclosing prefix is registered rather than only the immediate owner, so a call into a nested type still counts as reaching into the type around it — which is what the `startswith` version did.

File: static_analyzer/engine/adapters/java_adapter.py [hunk 5]
```diff
@@ -285,6 +285,11 @@ def edge_strategy(self) -> EdgeStrategy:
         """Use definition-based edges — JDTLS serializes references requests."""
         return EdgeStrategy.DEFINITIONS
 
+    @property
+    def expands_constructors(self) -> bool:
+        """JDTLS resolves ``new Dog()`` to the class, so the constructor needs adding."""
+        return True
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/adapters/java_adapter.py:291]: Real, but it predates this PR rather than arriving with it: on `main` the expansion also lives only in `_postprocess_edges`, and the incremental cross-boundary paths never fanned a class edge out to its constructors either. `class_to_ctors` has only ever had that one consumer.

So an incremental run omitting a constructor edge that a full rebuild emits is existing behaviour, not a regression here, and I would rather fix it where it belongs — in `_restore_inbound_edges_via_definitions` and `_add_outbound_edges_from_changed_files` — with its own test for the warm-start case. Filing it separately.

File: tests/static_analyzer/test_call_graph_builder.py [hunk 6]
```diff
@@ -359,6 +361,105 @@ def test_constructor_expansion_preserves_existing_ctor_call_sites(self):
...
+        st = builder._symbol_table
+        st._symbols["app.main"] = caller
+        st._symbols["app.Dog"] = cls
+        st._symbols["app.Dog(__init__)"] = ctor
+        file_key = str(Path("/project/app.py"))
+        st._file_symbols[file_key] = [caller, cls, ctor]
+        st._primary_file_symbols[file_key] = [caller, cls, ctor]
+        st.build_indices()
+
+        result = builder._postprocess_edges({("app.main", "app.Dog"): []})
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply tests/static_analyzer/test_call_graph_builder.py]: Agreed, the test could pass for the wrong reason. It now passes a real construction site, so the only thing that can keep the constructor edge out is the adapter opting out.

File: static_analyzer/engine/call_graph_builder.py [hunk 7]
```diff
@@ -303,18 +303,45 @@ def _postprocess_edges(self, edge_set: EdgeMap) -> EdgeMap:
...
+
+        # Sites where a caller reaches *inside* a class, indexed once: scanning the whole
+        # edge set per class edge is quadratic on a graph where every resolved callable also
+        # adds its parent-class edge.
+        member_sites: dict[tuple[str, str], set[tuple[str, int, int]]] = {}
+        for (edge_src, edge_dst), sites in edge_set.items():
+            # Every enclosing scope, so a call into a nested type also counts as reaching
+            # into the type around it. Sliced before the first "(" -- a parameter list
+            # carries dots of its own.
+            parts = edge_dst.split("(", 1)[0].split(".")
```
Comment (ivanmilevtues) [inline static_analyzer/engine/call_graph_builder.py]: this seems like something we should be doing with our [source_inspector.py](https://github.com/CodeBoarding/CodeBoarding/blob/56a0b9ae1685b453efd1685dbdef295fd5535426/static_analyzer/engine/source_inspector.py#L153)

After chatting with an agent we got to:
> Could `SourceInspector` expose whether each `CallSite` is a constructor invocation, and have constructor expansion use that directly? It already knows the relevant AST node type. Reconstructing containment by splitting `edge_dst` couples this logic to qualified-name formatting. If graph-based inference is necessary, we should at least traverse `SymbolInfo.owner_qualified_name` rather than parse the name.

**Review by ivanmilevtues — APPROVED** [review]
No real blockers.
I just saw reintroduction to custom parsing outside of the source_inspector (tree_sitter) it would be good if the thing can live there.

You can also merge as is if it is not a simple prompt fix.

File: static_analyzer/engine/call_graph_builder.py (hunk 7 above)
Comment (Svilen-Stefanov) [reply static_analyzer/engine/call_graph_builder.py]: Agreed on both counts, and the first one is the better fix — done in `0a34c63a`.

`SourceInspector` already knew every shape of construction it needed: `_CONSTRUCTOR_NODE_TYPES`, C#'s target-typed `new(...)` and `: base(...)`, an applied attribute, and `_call_target_node` already maps each of those to the exact position the definition query was made at. So it now answers the question directly:

```python
def is_construction_site(self, site: CallSite) -> bool:
```

built in the **same tree walk that already populates the usage index**, so it costs no extra parsing and rides the existing per-file cache. `_postprocess_edges` just asks. The `edge_dst` splitting is gone, and with it the whole graph-inference index — the second half of your suggestion turned out to be unnecessary once the AST answers it.

Two things worth flagging, because measuring it changed the answer.

**`Dog::new` needs telling from `Dog::speak`.** Both are `method_reference`, so the node type alone isn't enough; the predicate checks for a `new` child. Java's `super(...)` is also a call node rather than a creation node, so it's named explicitly.

**The AST alone was not sufficient, and I only found that by running it.** Swapping the heuristic for the predicate took the Java edge-case graph from 110 edges to 113, and the three extra ones were wrong *overloads*: `new Dog("Rex", 5)` is genuinely a construction site, so expansion fired and fanned the class edge out to every constructor, adding `Main.main -> Dog.Dog(String)` beside the correct `Dog(String, int)`. The old heuristic had suppressed that by accident, because jdtls had already resolved that site to the right constructor and that counted as \"reaching inside the class\".

So the rule is now: expand at a site the source says constructs, **and** where the server has not already named a constructor itself. That keeps the overload precision and is still AST-driven rather than name-driven.

Verified on the real engine: this PR alone leaves the graph **identical to \`main\`** — 94 references, 86 edges, same sets. Under #544 it gives 110 edges, 0 lost, 24 gained, 0 invented, which is exactly what CodeBoarding-tests#27 already pins, so that PR needed no change.

New coverage: `TestIsConstructionSite` in `test_source_inspector.py` (Java `new`, `super(...)`, `Dog::new` vs `Dog::speak`, an argument *inside* a `new` which must not count, C# `new(5)` and `: base(n)`), and `TestPostprocessEdges` now runs against a real Java file on disk instead of synthetic paths, so the AST is actually consulted.
